community/cypher/cypher-compiler-3.1/src/test/scala/org/neo4j/cypher/internal/compiler/v3_1/planner/QueryGraphTest.scala

```diff
@@ -20,34 +20,107 @@
+    val pattRel2 = PatternRelationship(r2, (m, c), BOTH, Seq.empty, SimplePatternLength)
+    val qg = QueryGraph(patternRelationships = Set(pattRel1, pattRel2))
+
+    qg.findRelationshipsEndingOn(x) shouldBe empty
+    qg.findRelationshipsEndingOn(n) should equal(Set(pattRel1))
+    qg.findRelationshipsEndingOn(m) should equal(Set(pattRel1, pattRel2))
+    qg.findRelationshipsEndingOn(c) should equal(Set(pattRel2))
+  }
+
+  test("finds shortest path starting from a single element with a single node in the QG") {
+    val qg = QueryGraph(patternNodes = Set(n))
+
+    qg.smallestGraphIncluding(Set(n)) should equal(Set(n))
+  }
+
+  test("finds shortest path starting from a single element with a single relationship in the QG") {
+    val r = PatternRelationship(r1, (n, m), BOTH, Seq.empty, SimplePatternLength)
+    val qg = QueryGraph(patternRelationships = Set(r), patternNodes = Set(n, m))
+
+    qg.smallestGraphIncluding(Set(n)) should equal(Set(n))
+  }
+
+  test("finds shortest path starting from two nodes with a single relationship in the QG") {
+    val r = PatternRelationship(r1, (n, m), BOTH, Seq.empty, SimplePatternLength)
+    val qg = QueryGraph(patternRelationships = Set(r), patternNodes = Set(n, m))
+
+    qg.smallestGraphIncluding(Set(n, m)) should equal(Set(n, m, r1))
+  }
+
+  test("finds shortest path starting from two nodes with a two relationships in the QG") {
+    val pattRel1 = PatternRelationship(r1, (n, m), BOTH, Seq.empty, SimplePatternLength)
+    val pattRel2 = PatternRelationship(r2, (m, c), BOTH, Seq.empty, SimplePatternLength)
+    val qg = QueryGraph(patternRelationships = Set(pattRel1, pattRel2), patternNodes = Set(n, m, c))
+
+    qg.smallestGraphIncluding(Set(n, m)) should equal(Set(n, m, r1))
+  }
+
+  test("finds shortest path starting from two nodes with a two relationships between the same nodes in the QG") {
+    val pattRel1 = PatternRelationship(r1, (n, m), BOTH, Seq.empty, SimplePatternLength)
+    val pattRel2 = PatternRelationship(r2, (n, m), BOTH, Seq.empty, SimplePatternLength)
+    val qg = QueryGraph(patternRelationships = Set(pattRel1, pattRel2), patternNodes = Set(n, m))
+
+    val result = qg.smallestGraphIncluding(Set(n, m))
+    result should contain(n)
+    result should contain(m)
+    result should contain oneOf (r1, r2)
+  }
+
+  test("finds shortest path starting from two nodes with an intermediate relationship in the QG") {
+    val pattRel1 = PatternRelationship(r1, (n, m), BOTH, Seq.empty, SimplePatternLength)
+    val pattRel2 = PatternRelationship(r2, (m, c), BOTH, Seq.empty, SimplePatternLength)
+    val qg = QueryGraph(patternRelationships = Set(pattRel1, pattRel2), patternNodes = Set(n, m, c))
+
+    qg.smallestGraphIncluding(Set(n, c)) should equal(
+      Set(n, m, c, r1, r2))
+  }
+
+  test("find smallest graph that connect three nodes") { // MATCH (x)<--(a)-->(b), (a)-->(c) RETURN DISTINCT a,b,c
```

Comment:
The query in the comment is great. It would be even greater if the variables had the same names :)